apostrophecms/apostrophe
# PRO-5875

## Summary

New password hashing strategy based on `scrypt`, to address customer concerns that the existing solution has become excessively expensive in terms of CPU time while still following an OWASP best practices recommendation and utilizing an implementation already present in Node.js core.

## What are the specific steps to test this change?

See the extensive test coverage. However, you can also verify the speed improvement:

* WITHOUT this branch, add a user with a password to your site (or just verify you have an existing user on your site).
* Switch to this branch (`npm link` is pretty broken these days, so use a git dependency and a real `npm install`).
* Log in again as that user. It'll take the usual while, but it'll also update the password hash to the new format so the next login is faster.
* Now log out and log in a second time. It'll be significantly faster.

Notes should be added to the `@apostrophecms/user` module documentation regarding the new `scrypt` option, but this does not have to happen this very minute, as the defaults are secure and greatly outperform the old `pbkdf2`-based implementation.

## What kind of change does this PR introduce?
*(Check at least one)*


## Make sure the PR fulfills these requirements:


If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**
